### ISSUE TITLE
Proposal: Test Guards

### ISSUE THREAD

**Issue by MovGP0** [xunit/xunit#1070]
When a problem occurs because of a code change, often a dozend, sometimes hundreds of tests throw an error. This makes it time consuming to trace and find the root cause. 

In an ideal case only the test that is closest to the error should cause an error, while dependent tests should just issue an warning. 

Therefore, there should be a mechanism that makes tests dependent on other tests, while the order of execution of the tests should be random. 

### Decarative Variant

There could be an attribute that makes a test dependent on the successful execution of another test.  

    public sealed class MyObjectTests
    {
        public sealed class Constructor
        {
            [Fact]
            public void CreateObject()
            {
                // when this fails, the test should cause an error 
                Assert.DoesNotThrow(() => new MyObject());
            }
        }
        
        public sealed class SomeProperty
        {
            [DependsOn(typeof(Constructor), nameof(Constructor.MyConstructorTest))]
            public void MyPropertyTest()
            {
                // when this throws, the test should cause an warning instead of an error 
                var testObject = new MyObject(); 
                // ... 
            }
        }
    }

*Disadvantages*

- the test result is preliminary until the parent test has run; it might not be shown to the user or change in an unepected way 
- it is clear that the test doesn"t cover the root cause 

### Imperative Variant

There could be an Assert-Statement that issues the warning. 

    public sealed class MyObjectTests
    {
        public sealed class Constructor
        {
            [Fact]
            public void CreateObject()
            {
                // when this fails, the test should cause an error 
                Assert.DoesNotThrow(() => new MyObject());
            }
        }
        
        public sealed class SomeProperty
        {
            public void MyPropertyTest()
            {
                MyObject testObject;
                try 
                {
                    testObject = new MyObject(); 
                }
                catch
                { 
                    Assert.Inconclusive("The test could not be initialized"); 
                    return; 
                } 
                if(testObject == null) 
                { 
                    Assert.Inconclusive("The test could not be initialized");
                    return;
                }
                // ... 
            }
        }
    }

*Disadvantages*

- there is the potential to introduce lots of business logic into the tests, which should not be there. 

*Advantages*

- this is what NUnit does 
- it is immediately clear that there is something wrong with the test
- it is clear that the test doesn"t cover the root cause 

### Requiring at least one assert

If the test throws an exception, but not calling assert at least once, it should issue a warning instead of an error. 

*Disadvantages*

- potentially breaks existing logic, where any exception causes the test to fail 

*Advantages*

- easy to implement, because xUnit only needs some special handling dependent on the exception type (AssertException = error; any other Exception = warning)

### Priority Attribute

When a test fails, errors from tests with lower priority are shown as warning. 

*Disadvantages*

- same as [Dependent] property
- because priority is a number, it forces tevelopers to update those numbers if tests are changing 

### Other Ideas

I am open to suggestions if there are some better ideas how to handle those scenarios.

**Comment by bradwilson** [comment]
We have explicitly gone out of our way to make test dependencies as difficult as possible, because we believe they are a significant source of programmer error. This is why we run all tests in a random order, in parallel against unrelated tests.

While this idea may have merit in a test framework focused on integration testing, that is not our core purpose. When faced with conflicts in design between things that are best for unit testing vs. integration testing, we always fall in favor of unit testing.

I would suggest that this should be implemented as an extension to xUnit.net rather than part of the core framework.

**Comment by MovGP0** [comment]
I was in fact thinking about UnitTesting and not ComponentTesting or IntegrationTesting. 

Sometimes the error is ie. in the constructor of an object, or in a complex setup to get the object into the proper state, and all other tests for the object under test fail because of that. This can be hundreds of tests sometimes. In NUnit it is helpful to exclude tests with `Assert.Inconclusive()` in such cases. 

I agree that this is a special case and should be located in an extension. However, the xUnit test runner would have to handle the `AssertInconclusiveException`.

**Comment by another-guy** [comment]
@MovGP0 I had (felt) this idea for many months but was unable to express it in my mind.

And I completely agree that the problem does lay in Unit testing scenarios just like in the Integration testing ones. I wonder if the solution can be build on top of the existing functionality so that the core of xunit stays untouched and not polluted with side concepts.

I am not sure if anything like that exists, but there could be an test runner configuration, or some other (attribute-based) form of extensibility point exposed. IMO, it's really the test runner's responsibility to orchestrate all the tests and follow the custom rules if such are provided.

**Comment by MovGP0** [comment]
I guess a nice solution would be to have some method of marking the setup block: 

```csharp
[Fact]
public void SomeTest()
{
    // Arrange
    Begin.Arrange();
    // doing some elaborate setup 
    // if an exception happens here, it will be shown als warning 
    End.Arrange();

    // act

    // assert
    
}
```

**Comment by bradwilson** [comment]
Closed for age.
